avoid WebDriverException - java.net.ConnectException: Failed to connect to localhost...

2021-01-18T20:31:34.5805460Z org.openqa.selenium.WebDriverException: 
2021-01-18T20:31:34.5810490Z java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:27231
2021-01-18T20:31:34.5813570Z Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:17:03'
2021-01-18T20:31:34.5816540Z System info: host: 'Mac-1611000679179.local', ip: '10.40.0.39', os.name: 'Mac OS X', os.arch: 'x86_64', os.version: '10.15.7', java.version: '11.0.9'
2021-01-18T20:31:34.5818210Z Driver info: driver.version: SafariDriver
2021-01-18T20:31:34.5820160Z Caused by: java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:27231
2021-01-18T20:31:34.5821960Z Caused by: java.net.ConnectException: Connection refused (Connection refused)

--- a/primefaces-selenium-core/src/main/java/org/primefaces/extensions/selenium/spi/WebDriverProvider.java
+++ b/primefaces-selenium-core/src/main/java/org/primefaces/extensions/selenium/spi/WebDriverProvider.java
@@ -23,6 +23,7 @@ package org.primefaces.extensions.selenium.spi;
 
 import org.openqa.selenium.Dimension;
 import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebDriverException;
 import org.openqa.selenium.support.events.EventFiringWebDriver;
 import org.primefaces.extensions.selenium.PrimeSelenium;
 import org.primefaces.extensions.selenium.internal.ConfigProvider;
@@ -31,6 +32,8 @@ import org.primefaces.extensions.selenium.internal.OnloadScriptsEventListener;
 public class WebDriverProvider {
 
     private static final ThreadLocal<WebDriver> WEB_DRIVER = new ThreadLocal<>();
+
+    private static final int CREATE_WEBDRIVER_RETRIES = 3;
 
     public static void set(WebDriver driver) {
         WEB_DRIVER.set(driver);
@@ -44,8 +47,23 @@ public class WebDriverProvider {
         WebDriver driver = WEB_DRIVER.get();
         if (driver == null && create) {
             PrimeSeleniumAdapter adapter = ConfigProvider.getInstance().getAdapter();
+            int fails = 0;
 
-            driver = adapter.createWebDriver();
+            do {
+                /*
+                 * Avoid issues like 2021-01-18T20:31:34.5805460Z org.openqa.selenium.WebDriverException: 2021-01-18T20:31:34.5810490Z
+                 * java.net.ConnectException: Failed to connect to localhost/0:0:0:0:0:0:0:1:27231 which sometimes occur during Github Action jobs.
+                 */
+                try {
+                    driver = adapter.createWebDriver();
+                }
+                catch (WebDriverException ex) {
+                    fails++;
+                    if (fails >= CREATE_WEBDRIVER_RETRIES) {
+                        throw ex;
+                    }
+                }
+            } while (driver == null);
 
             /*
              * Define window-size for headless-mode. Selenium WebDriver-default seems to be 800x600. This causes issues with modern themes (eg Saga) which use

